test/core/storage/trie_pruner/trie_pruner_test.cpp

```diff
@@ -0,0 +1,635 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <gtest/gtest.h>
+
+#include <iostream>
+#include <random>
+
+#include "crypto/hasher/hasher_impl.hpp"
+#include "mock/core/application/app_configuration_mock.hpp"
+#include "mock/core/blockchain/block_tree_mock.hpp"
+#include "mock/core/storage/persistent_map_mock.hpp"
+#include "mock/core/storage/spaced_storage_mock.hpp"
+#include "mock/core/storage/trie/serialization/trie_serializer_mock.hpp"
+#include "mock/core/storage/trie/trie_storage_backend_mock.hpp"
+#include "mock/core/storage/write_batch_mock.hpp"
+#include "storage/trie/polkadot_trie/polkadot_trie_factory_impl.hpp"
+#include "storage/trie/polkadot_trie/polkadot_trie_impl.hpp"
+#include "storage/trie/serialization/polkadot_codec.hpp"
+#include "storage/trie/serialization/trie_serializer_impl.hpp"
+#include "storage/trie_pruner/impl/trie_pruner_impl.hpp"
+#include "testutil/literals.hpp"
+#include "testutil/outcome.hpp"
+#include "testutil/prepare_loggers.hpp"
+#include "testutil/storage/polkadot_trie_printer.hpp"
+
+using namespace kagome::storage;
+using namespace kagome::storage::trie_pruner;
+using namespace kagome::common;
+namespace primitives = kagome::primitives;
+namespace crypto = kagome::crypto;
+using kagome::primitives::BlockHash;
+using kagome::primitives::BlockHeader;
+using kagome::primitives::BlockInfo;
+using kagome::primitives::BlockNumber;
+using testing::_;
+using testing::A;
+using testing::An;
+using testing::Invoke;
+using testing::Return;
+using testing::ReturnRef;
+
+auto hash_from_str(std::string_view str) {
+  Hash256 hash;
+  std::copy_n(str.begin(), str.size(), hash.begin());
+  return hash;
+}
+
+auto hash_from_header(const BlockHeader &header) {
+  static crypto::HasherImpl hasher;
+  return hasher.blake2b_256(scale::encode(header).value());
+}
+
+class CodecMock : public trie::Codec {
+ public:
+  MOCK_METHOD(outcome::result<Buffer>,
+              encodeNode,
+              (const trie::TrieNode &opaque_node,
+               std::optional<trie::StateVersion> version,
+               const ChildVisitor &child_visitor),
+              (const, override));
+  MOCK_METHOD(outcome::result<std::shared_ptr<trie::TrieNode>>,
+              decodeNode,
+              (gsl::span<const uint8_t> encoded_data),
+              (const, override));
+  MOCK_METHOD(Buffer, merkleValue, (const BufferView &buf), (const, override));
+  MOCK_METHOD(outcome::result<Buffer>,
+              merkleValue,
+              (const trie::OpaqueTrieNode &opaque_node,
+               std::optional<trie::StateVersion> version,
+               const ChildVisitor &child_visitor),
+              (const, override));
+  MOCK_METHOD(bool, isMerkleHash, (const BufferView &buf), (const, override));
+  MOCK_METHOD(Hash256, hash256, (const BufferView &buf), (const, override));
+};
+
+class PolkadotTrieMock final : public trie::PolkadotTrie {
```

Comment:
Maybe extract it to separated file?

Reply:
This one is pretty specific to these tests, I'm not sure about that.